### DISCUSSION TITLE
Add AVIF Support

### DISCUSSION THREAD

**Discussion by JimBobSquarePants** [SixLabors/ImageSharp#1320] [Ideas]
Next generation royalty free image format. 

AVIF offers significant file size reduction for images compared with JPEG or WebP; ~50% savings compared to JPEG, and ~20% savings compared to WebP

https://www.ctrl.blog/entry/webp-avif-comparison.html

Specification.
https://aomediacodec.github.io/av1-avif/

Source BSD licensed.
https://github.com/AOMediaCodec/libavif

**Comment by 0xC0000054** [comment]
My Paint.NET AVIF plugin is another open-source implementation (MIT licensed).
https://github.com/0xC0000054/pdn-avif

The code that reads and writes the AVIF file format is C#, the AV1 decoding and encoding is handled by a C++ library that uses [libaom](https://aomedia.googlesource.com/aom/).
The C# code allocates large arrays in unmanaged memory to reduce the work for the garbage collector, to make this cross-platform you would have to change the `SafeProcessHeapBuffer` class to use one of the memory allocation methods in the `Marshal` class.

I did not use libavif because it currently terminates the process on out-of-memory conditions.
I also need some features that libavif does not currently support, such as splitting large images into an image grid.
Using an image grid massively reduces memory usage when encoding large images, the libaom encoder is very memory hungry when working with images at 4K+ resolutions (it can use over 2 GB of memory when encoding a 4K image).

**Reply by JimBobSquarePants** [reply]
Thanks for sharing this @0xC0000054 I'll have a good look!

Any implementation we create will be purely managed and will only use other libraries as reference. We're pretty good at managing memory usage so we'll be able to use existing tooling to keep it as low as possible. It's just a question of either finding the time to implement the format or someone willing to do it for us.

**Comment by bjarnef** [comment]
This would be really great to support. I have recently seen more tweet about this, e.g https://twitter.com/addyosmani/status/1403608002372980736 and latest versions of Chrome already support AVIF https://caniuse.com/avif

There is also a nice article here about AVIF: https://jakearchibald.com/2020/avif-has-landed/ 

Other resources:
https://web.dev/compress-images-avif/
https://squoosh.app/
https://avif.io/

**Reply by JimBobSquarePants** [reply]
Yeah, likely all superseded by jpegXL  though. We’ll have to weigh up the pros/cons and choose between the two. It’s a huge amount of work to replicate these things in managed code so must choose wisely!

**Reply by jonsneyers** [reply]
If you're considering to implement a new codec from spec, then likely both jxl and avif are quite a bit of work. Going by spec page count though: the av1 spec is about 700 pages, then add some for the avif container; the jxl spec is about 100 pages. No idea how good the correlation is between spec page count and actual implementation effort is though. For both, there are existing FOSS implementations available with a compatible license like BSD.

It might be wise to wait and see which one gets the most traction.

For jxl, should you be interested in starting a new implementation: all the spec editors can be reached on the jxl discord server (I am one of them). You can ask them any questions about the spec or how best to implement it (most of the libjxl devs are also there).

**Reply by JimBobSquarePants** [reply]
Oh wow, hello @jonsneyers ; big fan!

AVIF scares me a little, all container formats do tbh so it's more than likely we'll implement JXL first but as you say, waiting is wise. 

We'll definitely visit the Discord though when we get started, thanks for letting us know it exists!

**Reply by NexusXe** [reply]
It seems that AVIF has taken the lead here in terms of adoption. AV1 is becoming more prevalent with some hardware decoders coming out, and as far as I know these hardware decoders are also able to accelerate AVIF image decoding.

While I recognize it's more difficult, the fact that it's supported by default in many more areas than JXL is worth taking into consideration.

**Reply by br3aker** [reply]
@NexusXe 

AVIF was pushed a lot by Netflix AFAIK.
JpegXL is still being optimized so I'd expect it to become more popular in the future.

In the end of the day it's all about the exact person to implement a new codec. If somebody would take their time to implement AVIF - cool, ImageSharp would have AVIF codec implementation. I personally wanted to implement JpegXL for a really long time now but all my time for this lib is spent on existing JPEG codec.

**Comment by RobSchoenaker** [comment]
There appears to be a native C# implementation for Paint.Net. The code looks promising.
https://github.com/0xC0000054/pdn-avif

**Comment by martonx** [comment]
I'd like to upvote this idea. With latest browser generations AVIF is the next big image format after WEBP.

**Comment by dhubenko** [comment]
bump! we need avif

**Comment by nsnlookupjn** [comment]
Yes, how to make this a priority? all of a sudden, a significant proportion of creative works are using AVIF format exclusively. It will not be long before AVIF is more common than other image formats, at least on mainstream web sites. I would love to help if I could.

**Reply by JimBobSquarePants** [reply]
There an open PR. Please feel free to help review it.

**Comment by 0xC0000054** [comment]
There is a 3rd-party ImageSharp extension that adds HEIF/AVIF decoding support: https://www.nuget.org/packages/HeyRed.ImageSharp.Heif/

**Reply by Lovrenc** [reply]
It works well for some images, but seems to fail to decode some properly, like:

[sample.zip](https://github.com/user-attachments/files/16637008/sample.zip)

**Comment by ScottRFrost** [comment]
Came across this discussion while looking for a nice (hopefully fully managed / not shell exec) method to read / write AVIF in ASP.NET APIs.  We're getting a lot of request from this, and CanIUse.com is now reporting [94% availability](https://caniuse.com/avif) on it (basically everything other than IE).

It's beyond my skill set to implement directly, but hopefully someone can point me in the right direction.
